src/group_theory/exponent.lean

```diff
@@ -91,66 +112,190 @@ end
+  obtain ⟨k, hk : exponent G = p ^ _ * k⟩ := pow_padic_val_nat_dvd; try {apply_instance},
+  obtain ⟨t, ht⟩ := nat.exists_eq_succ_of_ne_zero h.ne',
+  refine ⟨g ^ k, _⟩,
+  rw ht,
+  apply order_of_eq_prime_pow,
+  { rwa [hk, mul_comm, ht, pow_succ', ←mul_assoc, nat.mul_div_cancel _ hp.pos, pow_mul] at hg },
+  { rw [←nat.succ_eq_add_one, ←ht, ←pow_mul, mul_comm, ←hk],
+    exact pow_exponent_eq_one g },
 end
 
+variable {G}
+
+@[to_additive] lemma exponent_ne_zero_iff_range_order_of_finite (h : ∀ g : G, 0 < order_of g) :
+  exponent G ≠ 0 ↔ (set.range (order_of : G → ℕ)).finite :=
+begin
+  refine ⟨λ he, _, λ he, _⟩,
+  { by_contra h,
+    obtain ⟨m, ⟨t, rfl⟩, het⟩ := set.infinite.exists_nat_lt h (exponent G),
+    exact pow_ne_one_of_lt_order_of' he het (pow_exponent_eq_one t) },
+  { lift (set.range order_of) to finset ℕ using he with t ht,
+    have htpos : 0 < t.prod id,
+    { refine finset.prod_pos (λ a ha, _),
+      rw [←finset.mem_coe, ht] at ha,
+      obtain ⟨k, rfl⟩ := ha,
+      exact h k },
+    suffices : exponent G ∣ t.prod id,
+    { intro h,
+      rw [h, zero_dvd_iff] at this,
+      exact htpos.ne' this },
+    refine exponent_dvd_of_forall_pow_eq_one _ _ (λ g, _),
+    rw [pow_eq_mod_order_of, nat.mod_eq_zero_of_dvd, pow_zero g],
+    apply finset.dvd_prod_of_mem,
+    rw [←finset.mem_coe, ht],
+    exact set.mem_range_self g },
+end
+
+@[to_additive] lemma exponent_eq_zero_iff_range_order_of_infinite (h : ∀ g : G, 0 < order_of g) :
+  exponent G = 0 ↔ (set.range (order_of : G → ℕ)).infinite :=
+have _ := exponent_ne_zero_iff_range_order_of_finite h,
+by rwa [ne.def, not_iff_comm, iff.comm] at this
+
+end monoid
+
+section left_cancel_monoid
+
+variable [left_cancel_monoid G]
+
 @[to_additive lcm_add_order_eq_exponent]
-lemma lcm_order_eq_exponent {H : Type u} [fintype H] [left_cancel_monoid H] :
-  (finset.univ : finset H).lcm order_of = exponent H :=
+lemma lcm_order_eq_exponent [fintype G] : (finset.univ : finset G).lcm order_of = exponent G :=
 begin
-  apply nat.dvd_antisymm (lcm_order_of_dvd_exponent H),
-  apply exponent_dvd_of_forall_pow_eq_one,
-  intro g,
-  have h : (order_of g) ∣ (finset.univ : finset H).lcm order_of,
-  { apply finset.dvd_lcm,
-    exact finset.mem_univ g },
-  cases h with m hm,
+  apply nat.dvd_antisymm (lcm_order_of_dvd_exponent G),
+  refine exponent_dvd_of_forall_pow_eq_one G _ (λ g, _),
+  obtain ⟨m, hm⟩ : order_of g ∣ finset.univ.lcm order_of := finset.dvd_lcm (finset.mem_univ g),
   rw [hm, pow_mul, pow_order_of_eq_one, one_pow]
 end
 
+@[to_additive]
+lemma exponent_ne_zero_of_fintype [fintype G] : exponent G ≠ 0 :=
+by simpa [←lcm_order_eq_exponent, finset.lcm_eq_zero_iff] using λ x, (order_of_pos x).ne'
+
+end left_cancel_monoid
+
+section comm_monoid
+
+variable [cancel_comm_monoid G]
+
+@[to_additive] lemma exponent_eq_Sup_order_of (h : ∀ g : G, 0 < order_of g) :
+  exponent G = Sup (set.range (order_of : G → ℕ)) :=
```

Comment:
We usually use `supr` for `Sup (set.range _)`. It's definitionally the same but comes with nice notation and lemmas.

Reply:
Ooh, I didn't know about this. Is there an alternative for `finset` as well, or should I just leave that as `max`?

Reply:
I just `rw supr` at the start; it seems to me that for proving things about it the `Sup` API is a bit better. But the `csupr` induction seemed cool; if I was rewriting the proof, I'd probably use that.